Assignment3/Breakout.java

```diff
@@ -0,0 +1,252 @@
+
+	/** Height of a brick */
+	private static final int BRICK_HEIGHT = 8;
+
+	/** Radius of the ball in pixels */
+	private static final int BALL_RADIUS = 10;
+
+	/** Offset of the top brick row from the top */
+	private static final int BRICK_Y_OFFSET = 70;
+
+	/** Number of turns */
+	private static int NTURNS = 3;
+
+	private static final double GRAVITY = 6;
+
+	private int remainingBricks = NBRICK_ROWS * NBRICKS_PER_ROW;
+	private int remainingLifeCount = NTURNS;
+	private int bottomLineOfCurrentRow;
+	private int currentBrickLeftSide;
+	private double vx;
+	private double vy = GRAVITY;
+	private static final int DELAY = 20;
+	private Color currentColor;
+	private GOval ball;
+	private GRect brick;
+	private GRect paddle;
+	private GObject collider;
+	private GLabel lostLife;
+	private GLabel middleText;
+	private RandomGenerator rgen = RandomGenerator.getInstance();
+
+	public void run() {
+		setup();
+		pause(4000);
+		while (remainingBricks > 0 && remainingLifeCount >= 0) {
+			playGame();
+			if (ball.getY() >= HEIGHT) {
+				userLostLife();
+				pause(4999);
+			}
+			if (remainingBricks == 0) {
+				userWon();
+			}
+		}
+	}
+
+	private void setup() {
+		showWelcomeText();
+		createBricks();
+		createPaddle();
+		createBall();
+		addMouseListeners();
+	}
+	
+	/**
+     * Controls building rows of bricks in different colors.
+     */
+
+	public void createBricks() {
+		bottomLineOfCurrentRow = BRICK_Y_OFFSET + BRICK_HEIGHT;
+		currentBrickLeftSide = BRICK_SEP;
+		for (int i = 1; i <= NBRICK_ROWS; i++) {
+			bottomLineOfCurrentRow = bottomLineOfCurrentRow + BRICK_HEIGHT + BRICK_SEP;
+			if (i <= 2) {
+				currentColor = Color.RED;
+			} else if (i <= 4) {
+				currentColor = Color.ORANGE;
+			} else if (i <= 6) {
+				currentColor = Color.YELLOW;
+			} else if (i <= 8) {
+				currentColor = Color.GREEN;
+			} else if (i <= 10) {
+				currentColor = Color.CYAN;
+			}
+			for (int j = 1; j <= NBRICKS_PER_ROW; j++) {
+
+				buildBrick(currentColor);
+				currentBrickLeftSide = currentBrickLeftSide + BRICK_WIDTH + BRICK_SEP;
+			}
+			currentBrickLeftSide = BRICK_SEP;
+		}
+	}
+	
+	/**
+     * Defines a unified format of a brick. Ensures that all bricks are the same size and can be colored.
+     */
+
+	public void buildBrick(Color c) {
+		brick = new GRect(currentBrickLeftSide, bottomLineOfCurrentRow, BRICK_WIDTH, BRICK_HEIGHT);
+		add(brick);
+		brick.setFilled(true);
+		brick.setColor(c);
+	}
+
+	public void createPaddle() {
+		paddle = new GRect(WIDTH / 2 - PADDLE_WIDTH / 2, HEIGHT - PADDLE_Y_OFFSET, PADDLE_WIDTH, PADDLE_HEIGHT);
+		add(paddle);
+		paddle.setFilled(true);
+	}
+
+	public void createBall() {
+		ball = new GOval(WIDTH / 2, HEIGHT / 2, BALL_RADIUS, BALL_RADIUS);
+		add(ball);
+		ball.setColor(Color.GREEN);
+		ball.setFilled(true);
+	}
+	
+	/**
+     * Defines the way that ball will be moving. 
+     */
+	
+	private void moveBall() {
+		vx = rgen.nextDouble(1.0, 3.0);
```

Comment:
Algorithm is incorrect. We want to have variable velocity when going in a straight direction. Only reflect when bouncing off of walls, bricks, or paddle.